src/core/iomgr/tcp_server_posix.c

```diff
@@ -464,7 +464,9 @@ grpc_tcp_listener *grpc_tcp_server_add_port(grpc_tcp_server *s,
     addr_len = sizeof(wild6);
     fd = grpc_create_dualstack_socket(addr, SOCK_STREAM, 0, &dsmode);
     sp = add_socket_to_server(s, fd, addr, addr_len);
-    allocated_port = sp->port;
```

Comment:
allocated_port doesn't really need to be a distinct variable; just check "sp != NULL && sp->port > 0" below.

Also, it doesn't make sense to assign "sp2 = sp" only in the case where port==0.  That should happen whenever two sockets are created, regardless of the port number.